### PR TITLE
Add support for global/project level webhooks

### PR DESCRIPTION
**Pull request by kzh**
This PR adds support for global and project scoped webhooks.

### REST Endpoints:
`<bitbucket url>/rest/sourcegraph-admin/1.0/webhook`

Listing:  
* `$ curl <bitbucket url>/rest/sourcegraph-admin/1.0/webhook`  

Creating:  
* `$ curl <bitbucket url>/rest/sourcegraph-admin/1.0/webhook -d 'scope=<scope>&identifier=<identifier>&events=<events>&external=<external>'`
  * scope: `global` | `project` | `repository`
  * identifier: name of project or repository (empty if scope is global)
  * events: comma separated list of [events](https://confluence.atlassian.com/bitbucketserver0516/event-payload-966061436.html?utm_campaign=in-app-help&utm_medium=in-app-help&utm_source=stash#Eventpayload-repositoryevents)
  * external: URL at which the events data will be sent  

Deleting:  
* `$ curl -X DELETE <bitbucket url>/rest/sourcegraph-admin/1.0/webhook -d 'id=<id>'`
* `$ curl -X DELETE <bitbucket url>/rest/sourcegraph-admin/1.0/webhook -d 'name=<name>'`

For authorization, include these flags: `-u admin:admin -H "X-Atlassian-Token: no-check"`.

The event payload aims to follow these JSON schemas:
- https://confluence.atlassian.com/bitbucketserver0516/event-payload-966061436.html?utm_campaign=in-app-help&utm_medium=in-app-help&utm_source=stash#Eventpayload-repositoryevents
- https://docs.atlassian.com/bitbucket-server/docs/5.16.0/reference/javascript/JSON.html

### Todo:
- [x] Persistence layer for storing Webhooks via ActiveObjects
- [x] [Expose REST endpoint](https://developer.atlassian.com/server/framework/atlassian-sdk/basics-of-exposing-rest-apis-via-plugins/) (`/webhooks`) for CRUDing Webhooks
- [x] [Serialize events to JSON](https://confluence.atlassian.com/bitbucketserver0516/event-payload-966061436.html?utm_campaign=in-app-help&utm_medium=in-app-help&utm_source=stash#Eventpayload-repositoryevents) via [google/gson](https://github.com/google/gson) and send out HTTP requests to external URLs
- [x] Support more event listeners for `PullRequestEvent`
- [x] Use project/repository ID instead of name slug for unique identifiers
- [x] Support webhook secret to validate HTTP requests
- [ ] Add unit tests
- [ ] Document available events and general behaviour

This is a part of sourcegraph/sourcegraph#6091.

### CODE REVIEW

File: src/main/java/com/sourcegraph/webhook/EventSerializer.java
Comment (kzh) [inline src/main/java/com/sourcegraph/webhook/EventSerializer.java:31]: I considered the option of using reflection to serialize these events, but I decided against it. The reflection approach would require unnecessary code complexity (and even more to allow flexibility in the JSON output). This isn't worth it for the few number of supported events. This decision might be revisited if we support hooking onto more events.

Ideally, the Bitbucket API would expose something to serialize these events, but it doesn't :(

**Comment by kzh** [comment]
This PR is ready for first round review. cc @nicksnyder @tsenart @unknwon

**Review by unknwon — APPROVED** [review]
Thanks for making this!

I am really not sure how should I review this... because all of them LGTM.




Maybe add more comments in the code explain things like:
1. How would the method be used/invoked.
2. What role a class/method is.

**Comment by tsenart** [comment]
@kzh: Would it be possible to remove the whitespace / formatting changes or put them in a separate commit?

File: src/main/java/com/sourcegraph/webhook/Dispatcher.java
Comment (tsenart) [inline src/main/java/com/sourcegraph/webhook/Dispatcher.java]: Would this be a better fit for a for loop?

`for (int retries = 5; retries > 0; retries--) { ... }`

File: src/main/java/com/sourcegraph/webhook/Dispatcher.java
Comment (tsenart) [inline src/main/java/com/sourcegraph/webhook/Dispatcher.java]: Please add a comment saying how long this is or extract to a var / constant.

File: src/main/java/com/sourcegraph/webhook/EventSerializer.java
Comment (tsenart) [inline src/main/java/com/sourcegraph/webhook/EventSerializer.java]: Can we make this RFC3339? https://golang.org/pkg/time/

File: src/main/java/com/sourcegraph/webhook/EventSerializer.java
Comment (tsenart) [inline src/main/java/com/sourcegraph/webhook/EventSerializer.java]: Java noob question: What is a `HashMap` doing here? Aren't we rendering an Object directly to a JSON String?

File: src/main/java/com/sourcegraph/webhook/EventSerializer.java
Comment (tsenart) [reply src/main/java/com/sourcegraph/webhook/EventSerializer.java:31]: The Bitbucket Webhook plugin I linked to in the RFC seems to have a way to solve [this problem generically](https://sourcegraph.com/github.com/Eernie/bitbucket-webhooks-plugin/-/blob/src/main/java/nl/topicus/bitbucket/api/PullRequestListener.java#L308). What's in the way for us to use a similar approach?

File: src/main/java/com/sourcegraph/webhook/WebhookListener.java
Comment (tsenart) [inline src/main/java/com/sourcegraph/webhook/WebhookListener.java:30]: This bit of code is quite non-obvious to me. It seems that we already have pieces in the `split` string array. Is this loop simply converting that type to a `List<String>`? Is there a simpler way to do that?

File: src/main/java/com/sourcegraph/webhook/WebhookListener.java
Comment (tsenart) [inline src/main/java/com/sourcegraph/webhook/WebhookListener.java:38]:
```suggestion
        register(PullRequestUpdatedEvent.class, "pr:updated");
```

File: src/main/java/com/sourcegraph/webhook/WebhookListener.java
Comment (tsenart) [inline src/main/java/com/sourcegraph/webhook/WebhookListener.java:39]:
```suggestion
        register(PullRequestReviewersUpdatedEvent.class, "pr:reviewers:updated");
```

File: src/main/java/com/sourcegraph/webhook/WebhookListener.java
Comment (tsenart) [inline src/main/java/com/sourcegraph/webhook/WebhookListener.java:42]:
```suggestion
        register(PullRequestParticipantReviewedEvent.class, "pr:review:needs_work");
```

File: src/main/java/com/sourcegraph/webhook/WebhookListener.java
Comment (tsenart) [inline src/main/java/com/sourcegraph/webhook/WebhookListener.java]: Debug output left behind?

File: src/main/java/com/sourcegraph/webhook/WebhookListener.java
Comment (tsenart) [inline src/main/java/com/sourcegraph/webhook/WebhookListener.java:39]:
```suggestion
        register(PullRequestReviewersUpdatedEvent.class, "pr:reviewers:updated");
```

Comment (tsenart) [reply src/main/java/com/sourcegraph/webhook/WebhookListener.java:39]: Ditto for others below.

File: src/main/java/com/sourcegraph/webhook/WebhookRouter.java
Comment (tsenart) [inline src/main/java/com/sourcegraph/webhook/WebhookRouter.java]: We'd rather use JSON to register webhooks. Can we make these all JSON endpoints?

File: src/main/java/com/sourcegraph/webhook/WebhookRouter.java
Comment (tsenart) [inline src/main/java/com/sourcegraph/webhook/WebhookRouter.java]: Is there a logger we should be using?

File: src/main/java/com/sourcegraph/webhook/WebhookRouter.java
Comment (tsenart) [inline src/main/java/com/sourcegraph/webhook/WebhookRouter.java]: We need a `secret` as well, so that events are sent with a HMAC header signature of the payload for authentication purposes. Otherwise we have no way to authenticate incoming webhook requests in Sourcegraph. See GitHub's implementation for reference: https://github.com/sourcegraph/sourcegraph/blob/master/enterprise/pkg/a8n/webhooks_test.go#L244

**Comment by kzh** [comment]
Thanks for the review! I'll ping to request the second round :)

File: src/main/java/com/sourcegraph/webhook/WebhookListener.java
Comment (kzh) [reply src/main/java/com/sourcegraph/webhook/WebhookListener.java:30]: The loop converts the input string into a List<String> containing all the prefixes. So `pr:comment:added` becomes `[pr, pr:comment, pr:comment:added]`.

This is so if a webhook is registered for `pr:comment`, it will handle all the sub events: `pr:comment:added`, `pr:comment:edited`, `pr:comment:deleted`. `pr` would handle all pull request events.

I will include a comment describing this method.

File: src/main/java/com/sourcegraph/webhook/WebhookListener.java
Comment (kzh) [reply src/main/java/com/sourcegraph/webhook/WebhookListener.java:38]: These keys are from the atlassian docs: [event paylods](https://confluence.atlassian.com/bitbucketserver0516/event-payload-966061436.html?utm_campaign=in-app-help&utm_medium=in-app-help&utm_source=stash#Eventpayload-repositoryevents).

File: src/main/java/com/sourcegraph/webhook/EventSerializer.java
Comment (kzh) [reply src/main/java/com/sourcegraph/webhook/EventSerializer.java]: The `HashMap` is for the `options` parameter of the `render` function. I think passing in a `null` will throw a null pointer exception, so I just pass in an empty one. 

I'm not exactly sure how this parameter works either since the docs don't say much 🤔https://docs.atlassian.com/bitbucket-server/javadoc/4.0.2/spi/reference/com/atlassian/bitbucket/json/JsonRenderer.html

File: src/main/java/com/sourcegraph/webhook/EventSerializer.java
Comment (kzh) [reply src/main/java/com/sourcegraph/webhook/EventSerializer.java:31]: JSON serializers like `gson` and `jackson` (the one they use) don't work on the native bitbucket event objects. Their approach was to create [their own classes](https://sourcegraph.com/github.com/Eernie/bitbucket-webhooks-plugin/-/blob/src/main/java/nl/topicus/bitbucket/model/Models.java) that mirror each native bitbucket event objects. So whenever they wanted to serialize, they would convert the native objects to one of their classes and run `jackson` on that.

I wanted to avoid creating all these extra classes, so in a way, I defined the class formats in this file through the various adapters. I have one more idea that I'm going to try.

File: src/main/java/com/sourcegraph/webhook/EventSerializer.java
Comment (mrnugget) [inline src/main/java/com/sourcegraph/webhook/EventSerializer.java:67]: I'm assuming these are all pull request related events, not just a subset, right?

**Review by mrnugget — COMMENTED** [review]
Leaving this comment here to say: Thorsten was here 😄 I don't know enough about Bitbucket Server's plugin architecture to spot foundational mistakes and my knowledge of Java is superficial, but, that being said, I can't spot obvious mistakes here.

I'm wondering whether we should add some developer/debug experience features such as logging of _succesful_ deliveries, sending a test even for a given PR?

File: src/main/java/com/sourcegraph/webhook/WebhookListener.java
Comment (tsenart) [reply src/main/java/com/sourcegraph/webhook/WebhookListener.java:30]: Cool. Additionally, instead of `pieces`, you could name the variable `prefixes` for added clarity.

File: src/main/java/com/sourcegraph/webhook/WebhookListener.java
Comment (tsenart) [reply src/main/java/com/sourcegraph/webhook/WebhookListener.java:38]: I see. Worth documenting that.

File: src/main/java/com/sourcegraph/webhook/EventSerializer.java
Comment (kzh) [reply src/main/java/com/sourcegraph/webhook/EventSerializer.java:67]: The exhaustive list of supported events is in `WebhookListener.java`. It should cover the vast majority (95%) of pull request events. I think the other 5% events might be irrelevant/obscure (ex. PullRequestRescopedEvent - fires when more files added to PR after approval), so I left them out for now.

The ones that are listed here in this static block are just a subset. They are here because they require additional JSON fields for the data from the specific event (in addition to the fields common to all pull request events).

File: src/main/java/com/sourcegraph/webhook/Dispatcher.java
Comment (nicksnyder) [inline src/main/java/com/sourcegraph/webhook/Dispatcher.java:26]: Please add class and method doc comments here and everywhere else.

File: src/main/java/com/sourcegraph/webhook/Dispatcher.java
Comment (nicksnyder) [inline src/main/java/com/sourcegraph/webhook/Dispatcher.java]:
```suggestion
```

File: src/main/java/com/sourcegraph/webhook/Dispatcher.java
Comment (tsenart) [inline src/main/java/com/sourcegraph/webhook/Dispatcher.java]: In order to know what to parse on the receiving end of the webhook, we need to know what kind of webhook it is. Can you set a header with the name of the webhook please? This is what the headers are named in GitHub: https://developer.github.com/webhooks/#delivery-headers

File: src/main/java/com/sourcegraph/webhook/Dispatcher.java
Comment (nicksnyder) [inline src/main/java/com/sourcegraph/webhook/Dispatcher.java]: Should execution continue if signing fails?

File: src/main/java/com/sourcegraph/webhook/WebhookRouter.java
Comment (tsenart) [inline src/main/java/com/sourcegraph/webhook/WebhookRouter.java:31]: We'll need some basic documentation on how to use this API (but that can come in a follow up PR).

Comment (tsenart) [reply src/main/java/com/sourcegraph/webhook/WebhookRouter.java:31]: Question that comes to mind: Are requests authenticated in the same way as the normal Bitbucket Server API?

File: src/main/java/com/sourcegraph/webhook/Dispatcher.java
Comment (nicksnyder) [inline src/main/java/com/sourcegraph/webhook/Dispatcher.java]: It is a bit weird to have the loop condition here if you are going to explicitly check it on line 78. Can you structure the code so you do one or the other (not both).

Related: MAX_RETRIES would be better named MAX_ATTEMPTS as it stands because the first attempt isn't really a "retry".

File: src/main/java/com/sourcegraph/webhook/Dispatcher.java
Comment (nicksnyder) [inline src/main/java/com/sourcegraph/webhook/Dispatcher.java]: Is response only null if an exception happens? If so, would it be simpler to include this in the try block so you can eliminate the null check?

**Review by tsenart — APPROVED** [review]
Approving to unblock this PR but left some additional feedback, some of which can definitely be addressed in follow up PRs.

A more general question I have: How has all of this been tested? I think we'd truly benefit from at least one high level system level integration test to make our future lives easier in changing this code.

File: src/main/java/com/sourcegraph/webhook/EventSerializer.java
Comment (nicksnyder) [reply src/main/java/com/sourcegraph/webhook/EventSerializer.java]: I had a similar question. Can you assign this to a variable so it is clearer?

```suggestion
        HashMap<> options = new HashMap<>(); 
        String raw = renderer.render(o, options);
```

Do we really have to indirect through a JSON string? There is no method to go directly from a regular object to a JsonElement?

File: src/main/java/com/sourcegraph/webhook/EventSerializer.java
Comment (nicksnyder) [inline src/main/java/com/sourcegraph/webhook/EventSerializer.java]: I would invert the control flow here because it is clearer and saves one line of code.

```java
if (adapter != null) {
    adapter.apply(payload, event);
}
return payload;
```

File: src/main/java/com/sourcegraph/webhook/registry/Webhook.java
Comment (nicksnyder) [inline src/main/java/com/sourcegraph/webhook/registry/Webhook.java]: Prefer newline at end of files (here and other files).

**Review by nicksnyder — COMMENTED** [review]
No need to block on my review. I glanced through the code and here are some notes. My biggest ask is to add a lot of documentation.

File: src/main/java/com/sourcegraph/webhook/Dispatcher.java
Comment (kzh) [reply src/main/java/com/sourcegraph/webhook/Dispatcher.java]: Thanks for the catch!

File: src/main/java/com/sourcegraph/webhook/EventSerializer.java
Comment (kzh) [reply src/main/java/com/sourcegraph/webhook/EventSerializer.java]: The bitbucket API seems to only expose the `render` function which converts bitbucket objects (PullRequest, User, Ref, etc) directly to a JSON string, so there isn't much flexibility.

The other approach would be manually serializing these objects like how events are currently being handled to avoid incurring this redundant operation.

File: src/main/java/com/sourcegraph/webhook/WebhookRouter.java
Comment (kzh) [reply src/main/java/com/sourcegraph/webhook/WebhookRouter.java:31]: Yep!

```
Authentication
Any authentication that works against Bitbucket will work against the REST API. The preferred authentication methods are HTTP Basic (when using SSL) and OAuth. Other supported methods include: HTTP Cookies and Trusted Applications.

You can find OAuth code samples in several programming languages at bitbucket.org/atlassian_tutorial/atlassian-oauth-examples.

The log-in page uses cookie-based authentication, so if you are using Bitbucket in a browser you can call REST from JavaScript on the page and rely on the authentication that the browser has established.
```
(https://docs.atlassian.com/bitbucket-server/rest/5.16.0/bitbucket-rest.html)

Comment (tsenart) [reply src/main/java/com/sourcegraph/webhook/WebhookRouter.java:31]: This talks about the REST API. Does this plugin "extend" the same REST API? Or is it served under a different path? I guess all of these usage questions need to be documented.

File: src/main/java/com/sourcegraph/webhook/Dispatcher.java
Comment (tsenart) [inline src/main/java/com/sourcegraph/webhook/Dispatcher.java]: What on earth is this :D

Comment (kzh) [reply src/main/java/com/sourcegraph/webhook/Dispatcher.java]: I think I was testing out IntelliJ type autocomplete to see if there were any useful classes to use and it auto-imported 😅It's removed from the latest commit.

**Comment by tsenart** [comment]
@kzh: Can you please summarise the current state of this PR? What's left to do and what's your expected timeline?

File: src/main/java/com/sourcegraph/webhook/WebhookRouter.java
Comment (kzh) [reply src/main/java/com/sourcegraph/webhook/WebhookRouter.java:31]: Yes, I believe the plugin extend the native bitbucket REST API. They're both served under the `/rest/` path which requires authentication (else 401s).

**Comment by tsenart** [comment]
@christinaforney: By the way, since this plugin isn't tied to our release cycle, there isn't any hard requirement for us to ship this at the same time as 3.10. We have to implement support in Sourcegraph for receiving and processing these webhooks in 3.11 anyways, so no rush necessary here.

**Comment by kzh** [comment]
This PR is good to merge. Everything is fully functioning. I intend to include the flushed out docs (`README.md` for webhooks) + unit/integration tests + comments in a followup PR.

> A more general question I have: How has all of this been tested? I think we'd truly benefit from at least one high level system level integration test to make our future lives easier in changing this code.

So far, the tests have been manual. I spun up a small webserver that dumps incoming http requests:
```go
package main

import (
	"log"
	"net/http"
	"net/http/httputil"
)

func main() {
	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		dump, err := httputil.DumpRequest(r, true)
		if err != nil {
			log.Println(err)
		} else {
			log.Println(string(dump))
		}
	})
	log.Fatal(http.ListenAndServe(":80", nil))
}
```

I registered a webhook that listens to all PR events and sends it to the webserver.
`[{"id":1,"name":"test","scope":"global","identifier":"","events":["pr"],"endpoint":"<endpoint>","secret":"secret"}]`


This is the output on the webserver when I trigger an event (ex. merging a pr).
Screen Shot 2019-11-13 at 6 21 57 AM

**Comment by tsenart** [comment]
> This PR is good to merge. Everything is fully functioning. I intend to include the docs (README.md for webhooks) + unit/integration tests in a followup PR.

Great. Will that follow up PR also include Java docs that @nicksnyder asked for? If so, let's merge this!

**Comment by tsenart** [comment]
@kzh: I'd share that screenshot in our #progress channel! Thank you for working on this 🙇

**Pull request merged**
